### PR TITLE
Add missing Sublime/Atom commands

### PR DESCRIPTION
**Pull request by rebornix**
Add missing commands from Sublime/Atom per #3776 .

TODO: 
- [x] Join lines
- [x] Transpose
- [x] Upper/Lower case
- [x] Paste and format
- [x] Test cases

Discussed with Kai and Alex, our `contrib` part of code is used as internal extension so it's Okay to put these missing commands into our code base. However while implementing them in core, I found that there is one new issue: key bindings conflicts. Will have a discuss with Alex and @waderyan on Monday.

### CODE REVIEW

**Comment by wprater** [comment]
is paste and indent a larger feature as part of the formatting package? or something you'll add independently of this?

Indenting to he nearest line is not a practical solution, correct?

https://github.com/Microsoft/vscode/issues/4039

**Comment by rebornix** [comment]
Add test cases for these commands. Please note that we don't use the name `paste and indent` as that's reflecting what it really is, instead we use `paste and format`: paste the content and make sure the pasted content is formatted correctly.

Besides, the formatting of the content is decided by format provider. @wprater .

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 1]
```diff
@@ -6,10 +6,16 @@
 
 import * as nls from 'vs/nls';
 import { KeyCode, KeyMod, KeyChord } from 'vs/base/common/keyCodes';
+import { TPromise } from 'vs/base/common/winjs.base';
+import { isFalsyOrEmpty } from 'vs/base/common/arrays';
 import { SortLinesCommand } from 'vs/editor/contrib/linesOperations/common/sortLinesCommand';
+import { getDocumentRangeFormattingEdits } from 'vs/editor/contrib/format/common/format';
+import { EditOperationsCommand } from 'vs/editor/contrib/format/common//formatCommand';
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: two slashes here -> vs/editor/contrib/format/common//formatCommand

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 2]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+			let lineText = model.getLineContent(i);
+			let firstNonWhitespaceIdx = model.getLineFirstNonWhitespaceColumn(i);
+
+			if (firstNonWhitespaceIdx >= 1) {
+				let insertSpace = true;
+				if (trimmedLinesContent === '') {
+					insertSpace = false;
+				}
+
+				if (trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === ' ' ||
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: trimmedLinesContent.length could be 0 here, leading to a call of `trimmedLinesContent.charAt(-1)`. 

You could use if (insertSpace && ....)

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 3]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+					trimmedLinesContent.charAt(trimmedLinesContent.length - 1) === '\t') {
+					insertSpace = false;
+					trimmedLinesContent = trimmedLinesContent.replace(/[\s\uFEFF\xA0]+$/g, ' ');
+				}
+
+				let lineTextWithoutIndent = lineText.substr(firstNonWhitespaceIdx - 1);
+
+				if (lineTextWithoutIndent.charAt(0) === ')') {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Why are we looking for `)` in the line text here ? i.e. this looks weird to me, why not for `]`. I suggest to remove this unless it is part of some weird join lines spec.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 4]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+					new Selection(cursor.lineNumber, cursor.column + 1, cursor.lineNumber, cursor.column + 1)));
+			}
+		}
+
+		editor.executeCommands(this.id, commands);
+	}
+}
+
+@editorAction
+class PasteAndFormatAction extends EditorAction {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: :-1: IMHO on this action.

I would prefer to have an `editor.formatOnPaste` which in my opinion would align more with our current approach to automatic formatting (same as `formatOnType` and `formatOnSave`). @jrieken what do you think?

This one also assumes it knows how paste is implemented in the cursor, but in case of multicursor paste touches multiple ranges, not just one. i.e. `pastedContentRange` is incorrect in case of multicursor paste.

I suggest to wait with this action and consider implementing it in the format contribution, we are shipping since a long time without it, and there's no reason to rush it in.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 5]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+	public run(accessor: ServicesAccessor, editor: ICommonCodeEditor): void {
+		let selections = editor.getSelections();
+		let model = editor.getModel();
+		let commands: ICommand[] = [];
+
+		for (let i = 0, len = selections.length; i < len; i++) {
+			let selection = selections[i];
+			if (selection.isEmpty()) {
+				let cursor = selection.getStartPosition();
+				let word = model.getWordAtPosition(cursor);
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: `word` can be null here

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 6]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+				let wordRange = new Range(cursor.lineNumber, word.startColumn, cursor.lineNumber, word.endColumn);
+
+				if (wordRange !== undefined) {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: ???

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 7]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+	public run(accessor: ServicesAccessor, editor: ICommonCodeEditor): void {
+		let selections = editor.getSelections();
+		let model = editor.getModel();
+		let commands: ICommand[] = [];
+
+		for (let i = 0, len = selections.length; i < len; i++) {
+			let selection = selections[i];
+			if (selection.isEmpty()) {
+				let cursor = selection.getStartPosition();
+				let word = model.getWordAtPosition(cursor);
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: `word` can be null here. think of empty lines.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 8]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+				let wordRange = new Range(cursor.lineNumber, word.startColumn, cursor.lineNumber, word.endColumn);
+
+				if (wordRange !== undefined) {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: same ??

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts [hunk 9]
```diff
@@ -401,3 +407,267 @@ export class DeleteAllLeftAction extends EditorAction {
...
+				commands.push(new ReplaceCommandThatPreservesSelection(selection, text.toLocaleUpperCase(), selection));
+			}
+		}
+
+		editor.executeCommands(this.id, commands);
+	}
+}
+
+@editorAction
+export class LowerCaseAction extends EditorAction {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: I suggest to create an abstract `CaseAction` that has an abstract method `_modifyText` or whatever that can be overwritten to use `toLocaleUpperCase` and `toLocalLowerCase` to avoid code duplication.

File: src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts [hunk 10]
```diff
@@ -67,4 +67,118 @@ suite('Editor Contrib - Line Operations', () => {
...
+
+				editor.setSelection(new Selection(1, 12, 1, 12));
+				transposeAction.run(null, editor);
+				assert.equal(model.getLineContent(1), 'hell oworld', '005');
+				assert.deepEqual(editor.getSelection().toString(), new Selection(1, 12, 1, 12).toString(), '006');
+			}
+		);
+	});
+
+	test('toggle case', function () {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts]: add more tests, e.g. for empty lines or for the cursor in the middle of whitespace.

File: src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts [hunk 11]
```diff
@@ -67,4 +67,118 @@ suite('Editor Contrib - Line Operations', () => {
...
+				assert.deepEqual(editor.getSelection().toString(), new Selection(4, 2, 4, 8).toString(), '008');
+
+				editor.setSelection(new Selection(5, 1, 7, 3));
+				joinLinesAction.run(null, editor);
+				assert.equal(model.getLineContent(5), 'hello world', '009');
+				assert.deepEqual(editor.getSelection().toString(), new Selection(5, 1, 5, 3).toString(), '010');
+			});
+	});
+
+	test('transpose', function () {
```
Comment (alexandrudima) [inline src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts]: does it work for empty lines, or on empty files?

You can use scripts\test.bat --coverage and my awesome LCOV plugin to see your code coverage ;)

**Review by alexdima — CHANGES_REQUESTED** [review]
TL;DR I'd remove the formatAndPaste action for now, we can discuss more about it, I would prefer to have the responsibility of the format contribution to format on paste (with an option).

Otherwise, there's a typo that causes bundling errors, and some corner cases.

Nice job overall! :)

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 1 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Nice catch :)

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 2 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: I relied on `charAt(-1)` returning empty string but yes it's not a good one.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 3 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: The one I followed is Vim http://vimdoc.sourceforge.net/htmldoc/change.html#J but Sublime is doing this for all matching brackets. I think we can do the same, read matching brackets from language config. Right now I'll just remove it.

File: src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts (hunk 10 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts (hunk 11 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/test/common/linesOperations.test.ts]: They work well, added test cases for them as well.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 9 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Refactored. Now they should look similar to `abstract class AbstractFormatAction`. Thanks for pointing it out. I was too lazy at that moment.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 5 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 6 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 7 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 8 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 2 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

File: src/vs/editor/contrib/linesOperations/common/linesOperations.ts (hunk 1 above)
Comment (rebornix) [inline src/vs/editor/contrib/linesOperations/common/linesOperations.ts]: Addressed.

**Comment by rebornix** [comment]
@alexandrudima removed paste and indent action and addressed other comments.

**Comment by alexandrudima** [comment]
Nice! LGTM 👍

**Comment by thekalinga** [comment]
Can the shortcut be added to toggle case along with upper and lowercase, as this makes life easier to have just one shortcut to toggle case.

Which is what IDEA follows
